### PR TITLE
ci: update window vm to windows-2019

### PR DESCRIPTION
**Pull request by Keith-CY**

### CODE REVIEW

**Comment by Keith-CY** [comment]
```
2020-06-28T18:10:11.2985133Z Summary of all failing tests
2020-06-28T18:10:11.2985436Z FAIL tests/block-sync-render/indexer-connector.test.ts (12.99s)
2020-06-28T18:10:11.2985955Z   â— unit tests for IndexerConnector â€º #constructor â€º init with a node url and indexer folder path
2020-06-28T18:10:11.2986236Z 
2020-06-28T18:10:11.2986553Z     expect(jest.fn()).toHaveBeenCalledWith(...expected)
2020-06-28T18:10:11.2987136Z 
2020-06-28T18:10:11.2987820Z     Expected: "http://nodeurl:8114", "\\indexer\\data\\path"
2020-06-28T18:10:11.2988690Z     Received: "http://nodeurl:8114", "/indexer/data/path"
2020-06-28T18:10:11.2989119Z 
2020-06-28T18:10:11.2989551Z     Number of calls: 1
2020-06-28T18:10:11.2989902Z 
2020-06-28T18:10:11.3060414Z       109 |     });
2020-06-28T18:10:11.3061013Z       110 |     it('init with a node url and indexer folder path', () => {
2020-06-28T18:10:11.3061793Z     > 111 |       expect(stubbedIndexerConstructor).toHaveBeenCalledWith(nodeUrl, path.join(indexerFolderPath))
2020-06-28T18:10:11.3062488Z           |                                         ^
2020-06-28T18:10:11.3062895Z       112 |     });
2020-06-28T18:10:11.3063263Z       113 |   });
2020-06-28T18:10:11.3063688Z       114 |   describe('#connect', () => {
2020-06-28T18:10:11.3063965Z 
2020-06-28T18:10:11.3064565Z       at Object.<anonymous> (tests/block-sync-render/indexer-connector.test.ts:111:41)
2020-06-28T18:10:11.3064980Z 
2020-06-28T18:10:11.3065369Z FAIL tests/block-sync-render/queue.test.ts
2020-06-28T18:10:11.3065875Z   â— queue â€º #start â€º inits IndexerConnector
2020-06-28T18:10:11.3066255Z 
2020-06-28T18:10:11.3066622Z     expect(jest.fn()).toHaveBeenCalledWith(...expected)
2020-06-28T18:10:11.3067330Z 
2020-06-28T18:10:11.3067628Z     - Expected
2020-06-28T18:10:11.3076146Z     + Received
2020-06-28T18:10:11.3077636Z 
2020-06-28T18:10:11.3079221Z       [{"address": "ckt1qyqrdsefa43s6m882pcj53m4gdnj4k440axqswmu83", "addressIndex": 0, "addressType": 0, "balance": "", "blake160": "0xfakeblake160", "liveBalance": "", "path": "", "pendingBalance": "", "sentBalance": "", "txCount": 0, "version": "testnet", "walletId": ""}],
2020-06-28T18:10:11.3080575Z       "http://fakenode:8114",
2020-06-28T18:10:11.3080999Z     - StringMatching /test\/indexer_data/,
2020-06-28T18:10:11.3081665Z     + "C:\\Users\\VSSADM~1\\AppData\\Local\\Temp\\userData\\test\\indexer_data\\fakegenesisblockhash",
2020-06-28T18:10:11.3082109Z 
2020-06-28T18:10:11.3082482Z     Number of calls: 1
2020-06-28T18:10:11.3082801Z 
2020-06-28T18:10:11.3090441Z       171 |     });
2020-06-28T18:10:11.3090948Z       172 |     it('inits IndexerConnector', () => {
2020-06-28T18:10:11.3091515Z     > 173 |       expect(stubbedIndexerConnectorConstructor).toHaveBeenCalledWith(
2020-06-28T18:10:11.3092181Z           |                                                  ^
2020-06-28T18:10:11.3092686Z       174 |         addresses,
2020-06-28T18:10:11.3093073Z       175 |         fakeNodeUrl,
2020-06-28T18:10:11.3093604Z       176 |         expect.stringMatching(/test\/indexer_data/)
2020-06-28T18:10:11.3093933Z 
2020-06-28T18:10:11.3094426Z       at Object.<anonymous> (tests/block-sync-render/queue.test.ts:173:50)
2020-06-28T18:10:11.3094805Z 
2020-06-28T18:10:11.3095033Z 
2020-06-28T18:10:11.3095422Z Test Suites: 2 failed, 1 skipped, 42 passed, 44 of 45 total
2020-06-28T18:10:11.3096014Z Tests:       2 failed, 1 skipped, 419 passed, 422 total
2020-06-28T18:10:11.3096497Z Snapshots:   0 total
2020-06-28T18:10:11.3097285Z Time:        731.847s
2020-06-28T18:10:11.3105179Z Ran all test suites.
```

Tests failed @katat

**Comment by katat** [comment]
After fixed the tests to be compatible with windows env, the build failed with the following error message
```
[5/5] Building fresh packages...
error D:\a\1\s\node_modules\xxhash: Command failed.
info Visit https://yarnpkg.com/en/docs/cli/install for documentation about this command.
Exit code: 7
Command: node-gyp rebuild
Arguments: 
Directory: D:\a\1\s\node_modules\xxhash
Output:
gyp ERR! UNCAUGHT EXCEPTION 
gyp ERR! stack Error: spawn C:\Program Files (x86)\Microsoft Visual Studio\2019\Enterprise\MSBuild\15.0\Bin\MSBuild.exe ENOENT
gyp ERR! stack     at Process.ChildProcess._handle.onexit (internal/child_process.js:267:19)
gyp ERR! stack     at onErrorNT (internal/child_process.js:469:16)
gyp ERR! stack     at processTicksAndRejections (internal/process/task_queues.js:84:21)
gyp ERR! System Windows_NT 10.0.17763
gyp ERR! command "C:\\hostedtoolcache\\windows\\node\\12.18.1\\x64\\node.exe" "D:\\a\\1\\s\\node_modules\\libnpm\\node_modules\\node-gyp\\bin\\node-gyp.js" "rebuild"
gyp ERR! cwd D:\a\1\s\node_modules\xxhash
gyp ERR! node -v v12.18.1
gyp ERR! node-gyp -v v4.0.0
gyp ERR! This is a bug in `node-gyp`.
gyp ERR! Try to update node-gyp and file an Issue if it does not help:
gyp ERR!     <https://github.com/nodejs/node-gyp/issues>

```

**Comment by Keith-CY** [comment]
> @Keith-CY After fixed the tests to be compatible with windows env, the build failed with the following error message
> 
> ```
> [5/5] Building fresh packages...
> error D:\a\1\s\node_modules\xxhash: Command failed.
> info Visit https://yarnpkg.com/en/docs/cli/install for documentation about this command.
> Exit code: 7
> Command: node-gyp rebuild
> Arguments: 
> Directory: D:\a\1\s\node_modules\xxhash
> Output:
> gyp ERR! UNCAUGHT EXCEPTION 
> gyp ERR! stack Error: spawn C:\Program Files (x86)\Microsoft Visual Studio\2019\Enterprise\MSBuild\15.0\Bin\MSBuild.exe ENOENT
> gyp ERR! stack     at Process.ChildProcess._handle.onexit (internal/child_process.js:267:19)
> gyp ERR! stack     at onErrorNT (internal/child_process.js:469:16)
> gyp ERR! stack     at processTicksAndRejections (internal/process/task_queues.js:84:21)
> gyp ERR! System Windows_NT 10.0.17763
> gyp ERR! command "C:\\hostedtoolcache\\windows\\node\\12.18.1\\x64\\node.exe" "D:\\a\\1\\s\\node_modules\\libnpm\\node_modules\\node-gyp\\bin\\node-gyp.js" "rebuild"
> gyp ERR! cwd D:\a\1\s\node_modules\xxhash
> gyp ERR! node -v v12.18.1
> gyp ERR! node-gyp -v v4.0.0
> gyp ERR! This is a bug in `node-gyp`.
> gyp ERR! Try to update node-gyp and file an Issue if it does not help:
> gyp ERR!     <[https://github.com/nodejs/node-gyp/issues](https://github.com/nodejs/node-gyp/issues?q=is%3Aissue+is%3Aopen+sort%3Aupdated-desc)>
> ```

It seems to be an error from rebuilding lumos, have you ever seen this kind of exceptions before?

**Comment by katat** [comment]
I haven't seen this error before on windows. Is it due to some specific environment settings on this windows instance? Are there any differences in the setups between this windows VM and the previous one?

**Comment by Keith-CY** [comment]
> I haven't seen this error before on windows. Is it due to some specific environment settings on this windows instance? Are there any differences in the setups between this windows VM and the previous one?

It keeps the same workflow as before but appends some necessary configuration for vscode 2019.

I've picked this commit to the develop branch(https://github.com/nervosnetwork/neuron/pull/1744) and it worked well. So I think there must be something extra required by the refactor branch.

**Comment by Keith-CY** [comment]
I've created another pr which removes the config to see the effects.

**Comment by katat** [comment]
I see the pipeline script is using `npm` instead `yarn`. Could this be the cause?

**Comment by Keith-CY** [comment]
> I see the pipeline script is using `npm` instead `yarn`. Could this be the cause?

This [PR](https://github.com/nervosnetwork/neuron/pull/1746) use `yarn` instead of `npm` and it has the same error

**Comment by Keith-CY** [comment]
> I've created another pr which removes the config to see the effects.

This [PR](https://github.com/nervosnetwork/neuron/pull/1745) removes the config and it works well.

Now we have tried these:

- With `config` and merged into `develop` => OK
- Without `config` and merged into `develop` => OK
- With `config` and merged into `refactor` => Error
- Without `config` and merge into `refactor` => Error

**Comment by katat** [comment]
So the earliest error is related to LLVM on windows. Did you manage to install the LLVM dependency on windows?

**Comment by Keith-CY** [comment]
> So the earliest error is related to LLVM on windows. Did you manage to install the LLVM dependency on windows?

This [commit](https://github.com/nervosnetwork/neuron/pull/1739/files#diff-fec826feae04e51c0d94076385408bdcR70) enables llvm on windows

**Comment by Keith-CY** [comment]
As the log said, the exit code is 7, which meant there's an internal exception from `xxhash`, which is used by `lumos`

**Comment by katat** [comment]
So I took a look at the differences between these win19 PRs. The difference seems to be https://github.com/nervosnetwork/neuron/pull/1739/files#diff-fec826feae04e51c0d94076385408bdcR81-R82

**Comment by katat** [comment]
I just found this example: https://github.com/WebAssembly/wasi-libc/pull/78/files/8392142dce2ed8f9227e0e304b48b89689021794#diff-79a889188c51195ad1884f90e264f03cR21-R26
That seems to download the llvm binary and install via the pipeline script.

Maybe we can have a try with this on top of the original `pipeline.yml`

**Comment by katat** [comment]
The alternative way is to have a windows VM managed by us, and have it connected to the CI as an agent. Then we can manually install the LLVM dependency on it to see if it works.

On my windows VM, the only error throws when building lumos indexer is the LLVM error. Once the LLVM installed, the build works on my VM. So I guess it is mainly an issue around LLVM.

**Comment by Keith-CY** [comment]
> I just found this example: [WebAssembly/wasi-libc/pull/78/files/8392142dce2ed8f9227e0e304b48b89689021794#diff-79a889188c51195ad1884f90e264f03cR21-R26](https://github.com/WebAssembly/wasi-libc/pull/78/files/8392142dce2ed8f9227e0e304b48b89689021794#diff-79a889188c51195ad1884f90e264f03cR21-R26)
> That seems to download the llvm binary and install via the pipeline script.
> 
> Maybe we can have a try with this on top of the original `pipeline.yml`

I'll have a try now.

File: packages/neuron-wallet/tests/block-sync-render/indexer-connector.test.ts [hunk 1]
```diff
@@ -29,7 +29,7 @@ const connectIndexer = async (indexerConnector: IndexerConnector) => {
 
 describe('unit tests for IndexerConnector', () => {
   const nodeUrl = 'http://nodeurl:8114'
-  const indexerFolderPath = '/indexer/data/path'
+  const indexerFolderPath = process.platform === 'win32' ? '\\indexer\\data\\path' : '/indexer/data/path'
```
Comment (Keith-CY) [inline packages/neuron-wallet/tests/block-sync-render/indexer-connector.test.ts:32]: I'd like to know why not `path.join` here?

File: packages/neuron-wallet/tests/block-sync-render/queue.test.ts [hunk 2]
```diff
@@ -170,10 +170,17 @@ describe('queue', () => {
       await queue.start()
     });
     it('inits IndexerConnector', () => {
+      let expectPathRegex
+      if (process.platform === 'win32') {
+        expectPathRegex = expect.stringMatching(/test\\indexer_data/)
```
Comment (Keith-CY) [inline packages/neuron-wallet/tests/block-sync-render/queue.test.ts:175]: The `\\` and `/` could be replaced by `path.delimiter`

File: packages/neuron-wallet/tests/block-sync-render/indexer-connector.test.ts (hunk 1 above)
Comment (katat) [reply packages/neuron-wallet/tests/block-sync-render/indexer-connector.test.ts:32]: Good suggestion. I will update it using `path.join`. 👍

File: packages/neuron-wallet/tests/block-sync-render/queue.test.ts (hunk 2 above)
Comment (katat) [reply packages/neuron-wallet/tests/block-sync-render/queue.test.ts:175]: Not sure how to use `path.delimiter` in ragex here. Any suggestions?

Comment (Keith-CY) [reply packages/neuron-wallet/tests/block-sync-render/queue.test.ts:175]: `new RegExp('text' + path.delimiter + 'index_data')` should work

Comment (Keith-CY) [reply packages/neuron-wallet/tests/block-sync-render/queue.test.ts:175]: my fault, use `path.sep`

Comment (katat) [reply packages/neuron-wallet/tests/block-sync-render/queue.test.ts]: Thanks! It works.

**Comment by Keith-CY** [comment]
I'm not sure if it ok that there're some errors thrown during compiling lumos but the tests are all executed in this [PR](https://github.com/nervosnetwork/neuron/pull/1749)

**Comment by katat** [comment]
I am checking at the logs of the builds and trying to see if it really succeeded the installation of llvm. From this [log](https://dev.azure.com/nervosnetwork/neuron/_build/results?buildId=12715&view=logs&jobId=1dd1a9fa-5bd6-5a58-be75-4682264ffaf4&j=1dd1a9fa-5bd6-5a58-be75-4682264ffaf4&t=74958cbb-e240-5904-fa0e-42137502e9ee, it doesn't display any logs relevant to llvm), Is there a way on CI for us to dig into the actual logs generated from a specific pipeline script?

**Comment by katat** [comment]
> I'm not sure if it ok that there're some errors thrown during compiling lumos but the tests are all executed in this PR

If the build failed on windows, then the CI build should fail in the production environment.

**Comment by katat** [comment]
I create a new PR from this with a change to install llvm via `choco` on windows, and it works. https://dev.azure.com/nervosnetwork/neuron/_build/results?buildId=12733&view=logs&j=1dd1a9fa-5bd6-5a58-be75-4682264ffaf4&t=3f018128-e114-5377-a2ad-085d965e1d73

File: packages/neuron-wallet/tests/block-sync-render/queue.test.ts (hunk 2 above)
Comment (katat) [reply packages/neuron-wallet/tests/block-sync-render/queue.test.ts]: Actually the[ test failed ](https://dev.azure.com/nervosnetwork/neuron/_build/results?buildId=12734&view=logs&jobId=1dd1a9fa-5bd6-5a58-be75-4682264ffaf4&j=1dd1a9fa-5bd6-5a58-be75-4682264ffaf4&t=0fd07730-01dd-5646-2168-32456ff1e2aa)on window when using `path.sep`, which can't automatically make the path separator compatible on windows.

I revert the code for this test.

**Pull request merged**